// @ionic/angular/dist/directives/navigation/ion-tabs.js
import * as tslib_1 from "tslib";
import { Component, ContentChild, HostListener, ViewChild } from '@angular/core';
import { NavController } from '../../providers';
import { IonTabBar } from '../proxies';
import { IonRouterOutlet } from './ion-router-outlet';
var IonTabs = /** @class */ (function () {
    function IonTabs(navCtrl) {
        this.navCtrl = navCtrl;
    }
    IonTabs.prototype.onPageSelected = function (detail) {
        if (this.tabBar) {
            this.tabBar.selectedTab = detail.view.stackId;
        }
    };
    IonTabs.prototype.select = function (tab) {
        var selected = this.outlet.getActiveStackId() === tab;
        var href = this.outlet.tabsPrefix + "/" + tab;
        var url = selected
            ? href
            : this.outlet.getLastUrl(tab) || href;
        return this.navCtrl.navigateBack(url);
    };
    tslib_1.__decorate([
        ViewChild('outlet', { read: IonRouterOutlet }),
        tslib_1.__metadata("design:type", IonRouterOutlet)
    ], IonTabs.prototype, "outlet", void 0);
    tslib_1.__decorate([
        ContentChild(IonTabBar),
        tslib_1.__metadata("design:type", Object)
    ], IonTabs.prototype, "tabBar", void 0);
    tslib_1.__decorate([
        HostListener('ionRouterOutletActivated', ['$event.detail']),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", void 0)
    ], IonTabs.prototype, "onPageSelected", null);
    tslib_1.__decorate([
        HostListener('ionTabButtonClick', ['$event.detail.tab']),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [String]),
        tslib_1.__metadata("design:returntype", void 0)
    ], IonTabs.prototype, "select", null);
    IonTabs = tslib_1.__decorate([
        Component({
            selector: 'ion-tabs',
            template: "\n    <ng-content select=\"[slot=top]\"></ng-content>\n    <div class=\"tabs-inner\">\n      <ion-router-outlet #outlet tabs=\"true\"></ion-router-outlet>\n    </div>\n    <ng-content></ng-content>",
            styles: ["\n    :host {\n      display: flex;\n      position: absolute;\n      top: 0;\n      left: 0;\n      right: 0;\n      bottom: 0;\n\n      flex-direction: column;\n\n      width: 100%;\n      height: 100%;\n\n      contain: layout size style;\n      z-index: $z-index-page-container;\n    }\n    .tabs-inner {\n      position: relative;\n\n      flex: 1;\n\n      contain: layout size style;\n    }"
            ]
        }),
        tslib_1.__metadata("design:paramtypes", [NavController])
    ], IonTabs);
    return IonTabs;
}());
export { IonTabs };
