// @oktadev/schematics/src/add-auth/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@angular-devkit/core");
const schematics_1 = require("@angular-devkit/schematics");
const tasks_1 = require("@angular-devkit/schematics/tasks");
const schematics_utilities_1 = require("schematics-utilities");
function addPackageJsonDependencies(framework, options) {
    return (host, context) => {
        const dependencies = [];
        if (framework === exports.ANGULAR) {
            dependencies.push({ type: schematics_utilities_1.NodeDependencyType.Default, version: '~1.2.1', name: '@okta/okta-angular' });
        }
        else if (framework === exports.REACT || framework === exports.REACT_TS) {
            dependencies.push({ type: schematics_utilities_1.NodeDependencyType.Default, version: '~1.2.0', name: '@okta/okta-react' });
            dependencies.push({ type: schematics_utilities_1.NodeDependencyType.Default, version: '~5.0.0', name: 'react-router-dom' });
            if (framework === exports.REACT_TS) {
                dependencies.push({ type: schematics_utilities_1.NodeDependencyType.Default, version: '~4.3.2', name: '@types/react-router-dom' });
            }
        }
        else if (framework === exports.VUE || framework == exports.VUE_TS) {
            dependencies.push({ type: schematics_utilities_1.NodeDependencyType.Default, version: '~1.1.0', name: '@okta/okta-vue' });
            if (framework === exports.VUE_TS) {
                dependencies.push({ type: schematics_utilities_1.NodeDependencyType.Dev, version: '~1.0.2', name: '@types/okta__okta-vue' });
            }
        }
        else if (framework === exports.IONIC_ANGULAR) {
            dependencies.push({ type: schematics_utilities_1.NodeDependencyType.Default, version: '0.3.5', name: 'ionic-appauth' });
            dependencies.push({ type: schematics_utilities_1.NodeDependencyType.Default, version: '^2.2.0', name: '@ionic/storage' });
            if (options.platform === 'capacitor') {
                dependencies.push({ type: schematics_utilities_1.NodeDependencyType.Default, version: '^3.0.2', name: 'cordova-plugin-secure-storage' });
                dependencies.push({ type: schematics_utilities_1.NodeDependencyType.Default, version: '^2.0.9', name: 'cordova-plugin-advanced-http' });
                dependencies.push({ type: schematics_utilities_1.NodeDependencyType.Default, version: '^5.4.0', name: '@ionic-native/http' });
            }
        }
        dependencies.forEach(dependency => {
            schematics_utilities_1.addPackageJsonDependency(host, dependency);
            context.logger.log('info', `âœ…ï¸ Added '${dependency.name}' into ${dependency.type}`);
        });
        return host;
    };
}
function installPackageJsonDependencies() {
    return (host, context) => {
        context.addTask(new tasks_1.NodePackageInstallTask());
        context.logger.log('info', `ðŸ” Installing packages...`);
        return host;
    };
}
function getWorkspace(host) {
    const possibleFiles = ['/angular.json', '/.angular.json'];
    const path = possibleFiles.filter(path => host.exists(path))[0];
    const configBuffer = host.read(path);
    if (configBuffer === null) {
        throw new schematics_1.SchematicsException(`Could not find (${path})`);
    }
    const content = configBuffer.toString();
    return {
        path,
        workspace: core_1.parseJson(content, core_1.JsonParseMode.Loose),
    };
}
exports.ANGULAR = 'angular';
exports.REACT = 'react';
exports.REACT_TS = 'react-ts';
exports.VUE = 'vue';
exports.VUE_TS = 'vue-ts';
exports.IONIC_ANGULAR = 'ionic/angular';
function getFramework(host) {
    let possibleFiles = ['/package.json'];
    const path = possibleFiles.filter(path => host.exists(path))[0];
    const configBuffer = host.read(path);
    if (configBuffer === null) {
        throw new schematics_1.SchematicsException(`Could not find (${path})`);
    }
    else {
        const content = JSON.parse(configBuffer.toString());
        if (content.dependencies['@angular/core'] && !content.dependencies['@ionic/angular']) {
            return exports.ANGULAR;
        }
        else if (content.dependencies['react']) {
            if (content.dependencies['typescript']) {
                return exports.REACT_TS;
            }
            return exports.REACT;
        }
        else if (content.dependencies['vue']) {
            if (content.devDependencies['typescript']) {
                return exports.VUE_TS;
            }
            return exports.VUE;
        }
        else if (content.dependencies['@ionic/angular']) {
            return exports.IONIC_ANGULAR;
        }
        else {
            throw new schematics_1.SchematicsException('No supported frameworks found in your package.json!');
        }
    }
}
function addAuth(options) {
    return (host, context) => {
        // allow passing the framework in (for testing)
        let framework = options.framework;
        // if no framework defined, try to detect it
        if (!framework) {
            framework = getFramework(host);
        }
        let projectPath = './';
        if (framework === exports.ANGULAR) {
            const { workspace } = getWorkspace(host);
            if (!options.issuer) {
                throw new schematics_1.SchematicsException('You must specify an "issuer".');
            }
            const project = workspace.projects[Object.keys(workspace.projects)[0]];
            projectPath = project.root;
            if (!options.project) {
                options.project = Object.keys(workspace.projects)[0];
            }
            // add imports to app.module.ts
            schematics_utilities_1.addModuleImportToModule(host, projectPath + '/src/app/app.module.ts', 'AuthRoutingModule', './auth-routing.module');
        }
        if (framework == exports.IONIC_ANGULAR) {
            // add a package name from the issuer
            const parts = options.issuer.split('.');
            if (options.issuer.indexOf('.') === -1) {
                // hard-code a package name for localhost
                options.packageName = 'dev.localhost.ionic';
            }
            else {
                options.packageName =
                    parts[2].substring(0, parts[2].indexOf('/')) + '.'
                        + parts[1] + '.'
                        + parts[0].substring(parts[0].lastIndexOf('/') + 1);
            }
            // add cordova to package.json
            if (options.platform === 'cordova') {
                const content = host.read('./package.json');
                if (content) {
                    const pkgJson = JSON.parse(content.toString());
                    // save any pre-existing plugins
                    if (pkgJson.cordova && pkgJson.cordova.plugins) {
                        const existingPlugins = pkgJson.cordova.plugins;
                        pkgJson.cordova.plugins = Object.assign({}, cordovaNode(options.packageName).plugins, existingPlugins);
                        pkgJson.cordova.platforms = cordovaNode(options.packageName).platforms;
                    }
                    else {
                        pkgJson.cordova = cordovaNode(options.packageName);
                    }
                    host.overwrite('./package.json', JSON.stringify(pkgJson));
                }
            }
            // add imports to app.module.ts
            schematics_utilities_1.addModuleImportToModule(host, 'src/app/app.module.ts', 'HttpClientModule', '@angular/common/http');
            schematics_utilities_1.addModuleImportToModule(host, 'src/app/app.module.ts', 'AuthModule', './auth/auth.module');
            schematics_utilities_1.addModuleImportToModule(host, 'src/app/app.module.ts', 'IonicStorageModule.forRoot()', '@ionic/storage');
        }
        // Setup templates to add to the project
        const sourcePath = core_1.join(core_1.normalize(projectPath), 'src');
        const templatesPath = core_1.join(sourcePath, '');
        const templateSource = schematics_1.apply(schematics_1.url('./' + framework + '/src'), [
            schematics_1.template(Object.assign({}, options)),
            schematics_1.move(core_1.getSystemPath(templatesPath)),
            // fix for https://github.com/angular/angular-cli/issues/11337
            schematics_1.forEach((fileEntry) => {
                if (host.exists(fileEntry.path)) {
                    host.overwrite(fileEntry.path, fileEntry.content);
                }
                return fileEntry;
            }),
        ]);
        // Chain the rules and return
        return schematics_1.chain([
            options && options.skipPackageJson ? schematics_1.noop() : addPackageJsonDependencies(framework, options),
            options && options.skipPackageJson ? schematics_1.noop() : installPackageJsonDependencies(),
            schematics_1.mergeWith(templateSource, schematics_1.MergeStrategy.Overwrite),
        ])(host, context);
    };
}
exports.addAuth = addAuth;
function cordovaNode(packageName) {
    return {
        'plugins': {
            'cordova-plugin-advanced-http': {},
            'cordova-plugin-safariviewcontroller': {},
            'cordova-plugin-inappbrowser': {},
            'cordova-plugin-secure-storage': {},
            'cordova-plugin-customurlscheme': {
                'URL_SCHEME': packageName
            },
            'cordova-plugin-whitelist': {},
            'cordova-plugin-statusbar': {},
            'cordova-plugin-device': {},
            'cordova-plugin-splashscreen': {},
            'cordova-plugin-ionic-webview': {
                'ANDROID_SUPPORT_ANNOTATIONS_VERSION': '27.+'
            },
            'cordova-plugin-ionic-keyboard': {}
        },
        'platforms': [
            'android',
            'ios'
        ]
    };
}
exports.cordovaNode = cordovaNode;
