// ionic-appauth/lib/auth-action.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AuthActions;
(function (AuthActions) {
    AuthActions["Default"] = "Default";
    AuthActions["SignInSuccess"] = "Sign In Success";
    AuthActions["SignInFailed"] = "Sign In Failed";
    AuthActions["SignOutSuccess"] = "Sign Out Success";
    AuthActions["SignOutFailed"] = "Sign Out Failed";
    AuthActions["RefreshSuccess"] = "Refresh Success";
    AuthActions["RefreshFailed"] = "Refesh Failed";
    AuthActions["AutoSignInFailed"] = "Auto Sign In Failed";
    AuthActions["AuthSignInSuccess"] = "Auto Sign In Success";
})(AuthActions = exports.AuthActions || (exports.AuthActions = {}));
var AuthActionBuilder = /** @class */ (function () {
    function AuthActionBuilder() {
    }
    AuthActionBuilder.Default = function () {
        return {
            action: AuthActions.Default,
        };
    };
    AuthActionBuilder.SignOutSuccess = function () {
        return {
            action: AuthActions.SignOutSuccess,
        };
    };
    AuthActionBuilder.SignOutFailed = function () {
        return {
            action: AuthActions.SignOutFailed,
        };
    };
    AuthActionBuilder.RefreshSuccess = function (token) {
        return {
            action: AuthActions.RefreshSuccess,
            tokenResponse: token
        };
    };
    AuthActionBuilder.RefreshFailed = function () {
        return {
            action: AuthActions.RefreshFailed
        };
    };
    AuthActionBuilder.SignInSuccess = function (token) {
        return {
            action: AuthActions.SignInSuccess,
            tokenResponse: token
        };
    };
    AuthActionBuilder.SignInFailed = function () {
        return {
            action: AuthActions.AutoSignInFailed
        };
    };
    AuthActionBuilder.AutoSignInSuccess = function (token) {
        return {
            action: AuthActions.AuthSignInSuccess,
            tokenResponse: token
        };
    };
    AuthActionBuilder.AutoSignInFailed = function () {
        return {
            action: AuthActions.AutoSignInFailed,
        };
    };
    return AuthActionBuilder;
}());
exports.AuthActionBuilder = AuthActionBuilder;
